### DISCUSSION TITLE
Ban a spesific function/constructor

### DISCUSSION THREAD

**Discussion by VirtualTim** [checkstyle/checkstyle#12975] [Q&A] [answered]
Basically I'm looking at banning the specific Thread constructor [Thread](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#%3Cinit%3E(java.lang.Runnable))([Runnable](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Runnable.html) task), to force people to name their threads (ie: use [Thread](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#%3Cinit%3E(java.lang.Runnable,java.lang.String))([Runnable](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Runnable.html) task, [String](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/String.html) name)).

Currently I'm attempting to achieve this with regex, but it's pretty tricky as the runnable can be declared inline.
I was hoping that IllegalInstantiationCheck would achieve this, but that appears to ban all constructors, where I just want to ban one specific one.

Any ideas?

**Comment by nrmancuso** [comment]
@VirtualTim it sounds like [MatchXPath](https://checkstyle.sourceforge.io/config_coding.html#MatchXpath) could be perfect for your use case; we can help you to formulate an XPath expression. Please share an example of 

> but it's pretty tricky as the runnable can be declared inline

to make sure we can cover all cases.

**Reply by VirtualTim** [reply]
What I meant by inline Runnable is something like this:
```java
new Thread(new Runnable() {
  @Override
  public void run() {
    do_something();
  }
}, "thread name").start();
```
As opposed to:
```java
new Thread(do_something, "thread name").start();
```
or
```java
new Thread(() -> do_something(), "thread name").start();
```

So any suggestion on using MatchXpath to match any use of [Thread](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Thread.html#%3Cinit%3E(java.lang.Runnable))([Runnable](https://docs.oracle.com/en/java/javase/19/docs/api/java.base/java/lang/Runnable.html) task)?

I'm currently using the regex `\bThread\([^,]*\)`, but that generates false positives.

**Reply by romani** [reply]
Try to use forbiddenapis tool 

Example:
https://github.com/checkstyle/checkstyle/blob/c8ee5bab92c70fcf974aa449a255a780eef4a448/pom.xml#L1658
https://github.com/checkstyle/checkstyle/blob/c8ee5bab92c70fcf974aa449a255a780eef4a448/config/signatures.txt#L1

**Comment by nrmancuso** [comment, marked as answer]
@VirtualTim this should work for you:
```bash

➜  src cat Test.java 
public class Test {
    void inlineGood() {
        new Thread(new Runnable() {
            @Override
            public void run() {}
        }, "thread name").start();
    }

    void inlineBad() {
        new Thread(new Runnable() { // violation
            @Override
            public void run() {}
        }).start();
    }

    void goodThreads() {
        new Thread(m(), "thread name").start();
        new Thread(() -> m(), "thread name").start();
    }

    void badThreads() {
        new Thread(m()).start(); // violation
        new Thread(() -> m()).start(); // violation
    }

    Runnable m() {
        return System.out::println;
    }
}

➜  src cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MatchXpath">

      <property name="query" value="//LITERAL_NEW
                [./IDENT[@text='Thread'] and .[2 > count(./ELIST/*)]]"/>
      <message key="matchxpath.match"
              value="Thread should be named via 'Thread(task, name)' constructor."/>
    </module>
  </module>
</module>

➜  src java -jar checkstyle-10.7.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:10:9: Thread should be named via Thread(task, name) constructor. [MatchXpath]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:22:9: Thread should be named via Thread(task, name) constructor. [MatchXpath]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:23:9: Thread should be named via Thread(task, name) constructor. [MatchXpath]
Audit done.
Checkstyle ends with 3 errors.
```

**Reply by VirtualTim** [reply]
Awesome, thanks for that, I'll give it a shot.
Edit: Works great. Thanks @nrmancuso.
